storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3109,269 +3108,2022 @@ buf_block_t *fsp_get_latched_xdes_page(
+  if (!m_root)
+  {
+    mtr.commit();
+    m_index.lock.x_unlock();
+    return err;
+  }
+
+  m_root->page.fix();
+  mtr.release_last_page();
+  uint16_t level= btr_page_get_level(m_root->page.frame);
+  while (1)
+  {
+    err= defragment_level(level, &mtr, space_defrag);
+    DBUG_EXECUTE_IF("fail_after_level_defragment",
+                    if (m_index.table->id == 2 && level == 1)
+                      err= DB_CORRUPTION;);
+    if (err || !level)
+      break;
+    level--;
+  }
+  ut_ad(err == DB_SUCCESS || !mtr.has_modifications());
+  mtr.commit();
+  m_index.lock.x_unlock();
+  m_root->page.unfix();
+  return err;
+}
+
+/** check whether any user table exist in system tablespace
+@retval DB_SUCCESS_LOCKED_REC if user table exist
+@retval DB_SUCCESS if no user table exist
+@retval DB_CORRUPTION if any error encountered */
+static dberr_t user_tables_exists() noexcept
+{
+  mtr_t mtr;
+  btr_pcur_t pcur;
+  dberr_t err= DB_SUCCESS;
+  mtr.start();
+  for (const rec_t *rec= dict_startscan_system(&pcur, &mtr,
+                                               dict_sys.sys_tables);
+       rec; rec= dict_getnext_system(&pcur, &mtr))
+  {
+    const byte *field= nullptr;
+    ulint len= 0;
+    if (rec_get_deleted_flag(rec, 0))
+    {
+corrupt:
+      sql_print_error("InnoDB: Encountered corrupted record in SYS_TABLES");
+      err= DB_CORRUPTION;
+      goto func_exit;
+    }
+    field= rec_get_nth_field_old(rec, DICT_FLD__SYS_TABLES__SPACE, &len);
+    if (len != 4)
+      goto corrupt;
+    if (mach_read_from_4(field) != 0)
+      continue;
+    field= rec_get_nth_field_old(rec, DICT_FLD__SYS_TABLES__ID, &len);
+    if (len != 8)
+      goto corrupt;
+    if (!dict_sys.is_sys_table(mach_read_from_8(field)))
+    {
+      err= DB_SUCCESS_LOCKED_REC;
+      btr_pcur_close(&pcur);
+      goto func_exit;
+    }
+  }
+func_exit:
+  mtr.commit();
+  return err;
+}
+
+dberr_t fil_space_t::defragment() noexcept
+{
+  ut_ad(this == fil_system.sys_space);
+  dberr_t err= user_tables_exists();
+  if (err == DB_SUCCESS_LOCKED_REC)
+  {
+    sql_print_information(
+      "InnoDB: User table exists in the system tablespace."
+      "Please try to move the data from system tablespace "
+      "to separate tablespace before defragment the "
+      "system tablespace.");
+    return DB_SUCCESS;
+  } else if (err) { return err; }
+
+  SpaceDefragmenter defragmenter;
+  err= defragmenter.find_new_extents();
+  /* There is no free extent exist */
+  if (err == DB_SUCCESS_LOCKED_REC)
+    return DB_SUCCESS;
+
+  if (err == DB_SUCCESS)
+    err= defragmenter.defragment_system_tables();
+  return err;
+}
+
+/** Used during system tablespace truncation. Stores
+the "to be modified" extent descriptor page and its
+old page state */
+class fsp_xdes_old_page final
+{
+  std::vector<buf_block_t*> m_old_xdes_pages;
+  const uint32_t m_space;
+public:
+  fsp_xdes_old_page(uint32_t space):m_space(space) {}
+  ulint n_pages()
```

Comment:
This is missing `const noexpect` and the return type could be `uint32_t` to match the type of the variable that is being returned.